src/core/create-storage.ts

```diff
@@ -0,0 +1,181 @@
+}
+
+type Config<State> = Partial<
+  ConfigCreateStorage<State> & { key: string } & (
+      | ConfigAdapter
+      | ConfigAdapterFactory<any>
+    )
+>
+
+export function createStorage<State, Err = Error>(
+  ...configs: (string | Config<State> | undefined)[]
+): StorageHandles<State, Err> {
+  const config: Config<State> = {}
+  for (const cfg of configs) {
+    Object.assign(config, typeof cfg === 'string' ? { key: cfg } : cfg)
+  }
+
+  const {
+    adapter: adapterOrFactory,
+    context,
+    key: keyName,
+    keyPrefix = '',
+    contract,
+  } = config
+
+  if (!adapterOrFactory) {
+    throw Error('Adapter is not defined')
+  }
+  if (!keyName) {
+    throw Error('Key is not defined')
+  }
+
+  const adapter =
+    'factory' in adapterOrFactory ? adapterOrFactory(config) : adapterOrFactory
+
+  const key = keyName
+  const storage = getAreaStorage<State>(
+    adapter.keyArea || adapter,
+    keyPrefix + key
+  )
+
+  // const region = createNode()
+  // let disposable: (_: any) => void = () => {}
+  // const desist = () => disposable(clearNode(region))
+
+  const ctx = createStore<[any?]>(
+    [is.store(context) ? context.defaultState : undefined],
+    { serialize: 'ignore' }
+  )
+
+  const value = adapter<State>(keyPrefix + key, (x) => {
+    update(x)
+  })
+
+  // if (typeof value === 'function') {
+  //   disposable = value
+  // }
+
+  const fail = (
+    operation: 'get' | 'set' | 'validate',
+    error: unknown,
+    value?: any
+  ) =>
+    ({
+      key,
+      keyPrefix,
+      operation,
+      error,
+      value: typeof value === 'function' ? undefined : value, // hide internal "box" implementation
+    }) as Fail<Err>
+
+  const op = <T>(
+    operation: 'get' | 'set' | 'validate',
+    fn: (value: any, arg: any) => T,
+    value: any,
+    arg: any
+  ): T => {
+    try {
+      return fn(value, arg)
+    } catch (error) {
+      throw fail(operation, error, value)
+    }
+  }
+
+  const getFx = attach({
+    source: ctx,
+    effect([ref], raw?: void) {
+      const result = op('get', value.get, raw, ref) as any
+      return typeof result?.then === 'function'
+        ? Promise.resolve(result)
+            .then((result) => op('validate', validate, result, contract))
+            .catch((error) => {
+              throw fail('get', error, raw)
+            })
+        : op('validate', validate, result, contract)
+    },
+  }) as Effect<void, State, any> // as Effect<void, State, Fail<Err>>
+
+  const setFx = attach({
+    source: ctx,
+    effect([ref], state: State) {
+      const result = op('set', value.set, state, ref)
+      if (typeof result?.then === 'function') {
+        return Promise.resolve(result)
+          .then(() => undefined)
+          .catch((error) => {
+            throw fail('set', error, state)
+          })
+      }
+    },
+  }) as Effect<State, void, any> // as Effect<State, void, Fail<Err>>
+
+  const removeFx = attach({
+    mapParams: () => undefined as any,
```

Comment:
I'm afraid that using `undefined` as new storage value will break synchronization across different persisted entities... I thought about adding extra support from adapter, like, adding `removeItem` support for `localStorage` and `sessionStorage`